### PR TITLE
Add methods for listing and fetching container stats

### PR DESCRIPTION
**Pull request by theatrus**
This uses the previously unusued lib/stats.go code to return data
about container stats to the CRI API.

No data on filesystem layer usage is returned at this time.

Fixes one-half of #1248

Signed-off-by: Yann Ramin <atrus@stackworks.net>

**- Description for the changelog**

Add CRI API support for container statistics

### CODE REVIEW

**Comment by mheon** [comment]
@baude was working in this area of the code recently, and I think he found that the output of the stats code was questionable. The errors could also have been the frontend CLI we had to display the results, but I'd still be careful and make sure the results it's returning seem reasonable.

**Comment by theatrus** [comment]
Thanks for the word of caution - CPU was totally broken in the initial version (since it never recorded cpu time), but I'm going to take a deeper look at the reported numbers.

File: server/container_stats_list.go [hunk 1]
```diff
@@ -15,5 +15,32 @@ func (s *Server) ListContainerStats(ctx context.Context, req *pb.ListContainerSt
...
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var allStats []*pb.ContainerStats
+
+	for _, container := range containers.Containers {
+		stats, err := s.ContainerStats(ctx, &pb.ContainerStatsRequest{ContainerId: container.Id})
```
Comment (mrunalp) [inline server/container_stats_list.go]: Can we add a helper function to avoid the double look up?

Comment (theatrus) [reply server/container_stats_list.go]: Will do - I'l refactor out the ListContainers filter logic and ContainerStats.

Comment (theatrus) [reply server/container_stats_list.go]: I got completely derailed by a number of other totally unrelated tasks. I should be able to quickly refactor this by worst case tomorrow :)

Comment (mrunalp) [reply server/container_stats_list.go]: No worries. Thanks :)

File: server/container_list.go [hunk 2]
```diff
@@ -28,6 +28,45 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
 	return true
 }
 
+// filterContainerList applies a protobuf-defined filter to retrieve only intended containers. Not matching
+// the filter is not considered an error but will return an empty response.
+func (s *Server) filterContainerList(filter *pb.ContainerFilter) (ctrList []*oci.Container, err error) {
+
```
Comment (runcom) [inline server/container_list.go]: remove blank line

File: server/container_stats.go [hunk 3]
```diff
@@ -4,10 +4,33 @@ import (
...
 	"time"
 
+	"github.com/kubernetes-incubator/cri-o/lib"
+	"github.com/kubernetes-incubator/cri-o/oci"
 	"golang.org/x/net/context"
 	pb "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
 )
 
+func buildContainerStats(stats *lib.ContainerStats, container *oci.Container) *pb.ContainerStats {
+
```
Comment (runcom) [inline server/container_stats.go]: nit: remove blank

**Comment by TomSweeneyRedHat** [comment]
@theatrus looks like there are issues in the CI tests.  Are there any adjustments needed there or in the man pages for these changes?

**Comment by theatrus** [comment]
Need to dig into the integration failure (its at least consistent) - I haven't had a chance to re-run my code in cluster, and its likely I broke something.

**Comment by theatrus** [comment]
These e2e failures look like flakes:

```    Kubernetes e2e suite.[sig-scheduling] ResourceQuota should create a ResourceQuota and capture the life of a service.
    Kubernetes e2e suite.[sig-scheduling] ResourceQuota should create a ResourceQuota and capture the life of a persistent volume claim with a storage class. [sig-storage]
    Kubernetes e2e suite.[sig-scheduling] ResourceQuota should verify ResourceQuota with terminating scopes.```

File: server/container_list.go [hunk 4]
```diff
@@ -28,6 +28,46 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
...
+					ctrList = []*oci.Container{c}
+				} else {
+					ctrList = []*oci.Container{}
+				}
+
+			} else {
+				ctrList = []*oci.Container{c}
+			}
+		}
+	} else {
```
Comment (rhatdan) [inline server/container_list.go]: Change entire section from
```
+			if filter.PodSandboxId != "" {
+				if c.Sandbox() == filter.PodSandboxId {
+					ctrList = []*oci.Container{c}
+				} else {
+					ctrList = []*oci.Container{}
+				}
+
+			} else {
+				ctrList = []*oci.Container{c}
+			}
```
to
```
if filter.PodSandbox  == "" or if c.Sandbox() != filter.PodSandboxId {
        return []*oci.Container{}
}
return []*oci.Container{c}

File: server/container_list.go [hunk 5]
```diff
@@ -28,6 +28,46 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
...
+		if filter.PodSandboxId != "" {
+			pod := s.ContainerServer.GetSandbox(filter.PodSandboxId)
+			if pod == nil {
+				ctrList = []*oci.Container{}
```
Comment (rhatdan) [inline server/container_list.go]: return []*oci.Container{}

File: server/container_list.go [hunk 6]
```diff
@@ -28,6 +28,46 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
...
+			} else {
+				ctrList = pod.Containers().List()
```
Comment (rhatdan) [inline server/container_list.go]: return pod.Containers().List()

File: server/container_list.go [hunk 7]
```diff
@@ -28,6 +28,46 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
...
+			}
+		}
+	}
+	return
```
Comment (rhatdan) [inline server/container_list.go]: return origCtrList

Comment (theatrus) [reply server/container_list.go]: Thanks, changes made (and its much less spaghetti-like)

Comment (rhatdan) [reply server/container_list.go]: Still got a lint error.

File: server/container_list.go [hunk 8]
```diff
@@ -28,6 +28,38 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
...
+		if filter.PodSandboxId != "" {
+			pod := s.ContainerServer.GetSandbox(filter.PodSandboxId)
+			if pod == nil {
+				return []*oci.Container{}
+			} else {
+				return pod.Containers().List()
+			}
+		}
+	}
+	return origCtrList
```
Comment (TomSweeneyRedHat) [inline server/container_list.go]: It probably would be good to put up a debug statement before this return saying a filter list couldn't be determined or some such.

File: server/container_list.go [hunk 9]
```diff
@@ -28,6 +28,38 @@ func filterContainer(c *pb.Container, filter *pb.ContainerFilter) bool {
...
+				return []*oci.Container{}
+			}
+			return []*oci.Container{c}
+		}
+	} else {
+		if filter.PodSandboxId != "" {
+			pod := s.ContainerServer.GetSandbox(filter.PodSandboxId)
+			if pod == nil {
+				return []*oci.Container{}
+			} else {
```
Comment (TomSweeneyRedHat) [inline server/container_list.go]: lint doesn't like this else.   Just drop it and keep the return.

**Comment by theatrus** [comment]
Sorry for the test thrashing (teaches me to make 5 minute patches while on vacation, hi from an island somewhere!). 

I've replaced the inner logic block with the correct reduction, expressed as a switch statement for clarity (instead of nesting ifs). The original proposed reduction wasn't equivalent. This section really needs a unit test, but will require some mocking of the relevant objects as it stands today - I will follow up with a test PR here.

**Comment by theatrus** [comment]
@mrunalp - any last requests on this PR?

**Comment by mrunalp** [comment]
@theatrus Sorry for the delay. I'll make a final pass today. One thing we need to do is make the PR commits bisectable so that it builds at each commit so it might need some squashing. Thanks!

**Comment by theatrus** [comment]
Did a rebase step with the logical work bits in place. Will wait for CI to run over it again.

**Comment by mrunalp** [comment]
@theatrus Thanks! The code looks fine to me. We can merge after the tests pass.

**Pull request merged**